pkg/scalers/azure_eventhub_scaler.go

```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to get storage credentials: %s", err)
+	}
+
+	hub, err := GetEventHubClient(parsedMetadata.eventHubConnection)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get eventhub client: %s", err)
+	}
+
+	return &AzureEventHubScaler{
+		metadata:           parsedMetadata,
+		storageCredentials: cred,
+		client:             hub,
+	}, nil
+}
+
+// parseAzureEventHubMetadata parses metadata
+func parseAzureEventHubMetadata(metadata, resolvedEnv map[string]string) (*EventHubMetadata, error) {
+	meta := EventHubMetadata{}
+	meta.threshold = defaultEventHubMessageThreshold
+
+	if val, ok := metadata[thresholdMetricName]; ok {
+		threshold, err := strconv.ParseInt(val, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("Error parsing azure eventhub metadata %s: %s", thresholdMetricName, err)
+		}
+
+		meta.threshold = threshold
+	}
+
+	storageConnectionSetting := defaultStorageConnectionString
+	if val, ok := metadata["storageConnection"]; ok && val != "" {
+		storageConnectionSetting = val
+	} else {
+		return nil, fmt.Errorf("no storage connection setting given")
+	}
+
+	if val, ok := resolvedEnv[storageConnectionSetting]; ok {
+		meta.storageConnection = val
+	} else {
+		return nil, fmt.Errorf("no storage connection setting given")
+	}
+
+	eventHubConnectionSetting := defaultEventHubConnectionString
+	if val, ok := metadata["eventHubConnection"]; ok && val != "" {
+		eventHubConnectionSetting = val
+	} else {
+		return nil, fmt.Errorf("no event hub connection setting given")
+	}
+
+	if val, ok := resolvedEnv[eventHubConnectionSetting]; ok {
+		meta.eventHubConnection = val
+	} else {
+		return nil, fmt.Errorf("no event hub connection setting given")
+	}
+
+	meta.eventHubConsumerGroup = defaultEventHubConsumerGroup
+	if val, ok := metadata["consumerGroup"]; ok {
+		meta.eventHubConsumerGroup = val
+	}
+
+	return &meta, nil
+}
+
+//GetUnprocessedEventCountInPartition gets number of unprocessed events in a given partition
+func (scaler *AzureEventHubScaler) GetUnprocessedEventCountInPartition(ctx context.Context, partitionID string) (newEventCount int64, err error) {
+	partitionInfo, err := scaler.client.GetPartitionInformation(ctx, partitionID)
+	if err != nil {
+		return -1, fmt.Errorf("unable to get partition info: %s", err)
+	}
+
+	checkpoint, err := GetCheckpointFromBlobStorage(ctx, partitionID, *scaler.metadata)
+	if err != nil {
+		return -1, fmt.Errorf("unable to get checkpoint from storage: %s", err)
+	}
+
+	unprocessedEventCountInPartition := int64(0)
+
+	if checkpoint.SequenceNumber != partitionInfo.LastSequenceNumber {
+		if partitionInfo.LastSequenceNumber > checkpoint.SequenceNumber {
+			unprocessedEventCountInPartition = partitionInfo.LastSequenceNumber - checkpoint.SequenceNumber
+
+			return unprocessedEventCountInPartition, nil
+		} else {
+			unprocessedEventCountInPartition = (math.MaxInt64 - partitionInfo.LastSequenceNumber) + checkpoint.SequenceNumber
```

Comment:
why is this not just 0?

Reply:
according to EventHubActivationTask.cs, the partition is a circular buffer so it's possible for partitionInfo.LastSequenceNumber < checkpoint.SequenceNumber. It returns 0 only when the value returned is negative